src/sbd.h

```diff
@@ -210,3 +210,8 @@ void set_servant_health(enum pcmk_health state, int level, char const *format, .
 bool sbd_is_disk(struct servants_list_item *servant);
 bool sbd_is_pcmk(struct servants_list_item *servant);
 bool sbd_is_cluster(struct servants_list_item *servant);
+
+#define calculate_timeout_watchdog_warn(timeout_watchdog) \
+	(timeout_watchdog < 5 ? 2 : \
+	(timeout_watchdog * 3 <= INT32_MAX ? \
```

Comment:
Not saying we need that at all but if we have it
'timeout_watchdog < LONG_MAX / 3'
would be more what we should have here I guess.

Reply:
> Not saying we need that at all but if we have it
> 'timeout_watchdog < LONG_MAX / 3'
> would be more what we should have here I guess.

Right.

> As long as done in a macro casting timeout_watchdog to 'unsigned long' should be done as well.

AFAICT, it's alright since the passed value is never bigger than `unsigned long` and the result anyway fits into the `unsigned long timeout_watchdog_warn` that is being assigned.

Reply:
> 
> AFAICT, it's alright since the passed value is never bigger than `unsigned long` and the result anyway fits into the `unsigned long timeout_watchdog_warn` that is being assigned.

That is true but if you multiply it by 3 it might overflow and checking against LONG_MAX doesn't make sense anymore.
Which is why I had moved the 3 to the other side.
And in case of the initialization without casts or explicitly tagging the default as 'unsigned long' wouldn't it do int-calculations before a final cast to 'unsigned long' when assigning it to timeout_watchdog_warn?

Reply:
> That is true but if you multiply it by 3 it might overflow and checking against LONG_MAX doesn't make sense anymore.
> Which is why I had moved the 3 to the other side.

Right. That actually was the original way but somehow got messed up :P

> And in case of the initialization without casts or explicitly tagging the default as 'unsigned long' wouldn't it do int-calculations before a final cast to 'unsigned long' when assigning it to timeout_watchdog_warn?

I think preprocessor purely replaces the references the macros with the contents.  On initializing, still depending on the values of the operands,  if needed, it automatically does arithmetic conversion to the higher rank of data type.

Reply:
> > The thing is these would be changes of behavior and probably unnecessarily break existing setups. I'd prefer letting the sensible parts just take precedence.
> 
> We could try to just break setups that anyway didn't work probably.

Not necessarily true ... In case of inconsistent watchdog-timeout values between sysconfig and disk, so far the users have never needed to care about the consistency. They know the on-disk setting just takes precedence and it works.

About "stonith-watchdog-timeout is configured together with seeing any SBD-devices set via environment", an idea occurred to me is, if a sbd fencing resource is explicitly configured, pacemaker just ignore stonith-watchdog-timeout and de-register the internal watchdog-based device.

> I know that nobody is looking at error/warning-messages as long as it seems to be working but maybe at least as a first step we might introduce such messages.
> Or we pass it fully to high-level tooling.

That's at least a more friendly way.

Reply:
True, it's never wrong to be explicit.